File: apps/desktop/src/lib/library/ocrPage.ts
/**
 * Purpose: what a recognized page is on both editions, and how its pieces become the lines
 * that get indexed. The recognizer — the Rust `ocr_page` command on the desktop, the Worker
 * behind the same name in the browser — hands back the text lines it read, each with where it
 * sits on the page, and the blocks it found that are not running text: a table, read into
 * rows of cells, or a formula, read into LaTeX. Composing them is done here, once, in
 * TypeScript, so the two editions cannot drift on what a table looks like in a passage.
 *
 * The rules are few and deliberate. A block owns a line when most of the line sits inside it;
 * an owned line is not repeated, because the block already says it better. A formula box that
 * owns no whole line is a formula inside a sentence — the layout model finds those too — and
 * it is left to the text, where the OCR already read it as characters. A table is written as a
 * Markdown table because that is what the model reads and what keyword search indexes, and a
 * formula is written between `$$`, which is what the model writes back when it explains one.
 * Main exports: OcrBox, OcrLine, OcrBlock, OcrPageResult, composeRecognizedPage, tableToMarkdown.
 */

/** Page pixels, `x0,y0` top-left, `x1,y1` bottom-right. */
export interface OcrBox {
  x0: number;
  y0: number;
  x1: number;
  y1: number;
}

export interface OcrLine {
  text: string;
  /** Mean confidence of the characters read, 0 to 1. */
  score: number;
  box: OcrBox;
}

/** A table's `rows` are cells with spans already expanded to blanks; a formula's `latex`
 * carries no delimiters. */
export type OcrBlock =
  | { kind: "table"; box: OcrBox; rows: string[][] }
  | { kind: "formula"; box: OcrBox; latex: string };

export interface OcrPageResult {
  lines: OcrLine[];
  blocks: OcrBlock[];
}

/** How much of a line has to sit inside a block for the block to own it. */
export const OWNED_FRACTION = 0.7;

function area(box: OcrBox): number {
  return Math.max(0, box.x1 - box.x0) * Math.max(0, box.y1 - box.y0);
}

/** The share of `inner`'s area that lies inside `outer`, 0 to 1. */
export function insideFraction(inner: OcrBox, outer: OcrBox): number {
  const own = area(inner);
  if (own === 0) return 0;
  const overlap = area({
    x0: Math.max(inner.x0, outer.x0),
    y0: Math.max(inner.y0, outer.y0),
    x1: Math.min(inner.x1, outer.x1),
    y1: Math.min(inner.y1, outer.y1),
  });
  return overlap / own;
}

/** Top to bottom, left to right within a line — the same rule the recognizers use. */
function readingOrder(a: OcrBox, b: OcrBox): number {
  const tolerance = Math.max(10, 0.5 * Math.min(a.y1 - a.y0, b.y1 - b.y0));
  if (Math.abs(a.y0 - b.y0) < tolerance) return a.x0 - b.x0;
  return a.y0 - b.y0;
}

function escapeCell(text: string): string {
  return text.replace(/\|/g, "\\|").replace(/\s+/g, " ").trim();
}

/** Cells with text, as a share of all cells, below which a "table" is something else: the
 * layout model calls a periodic-table figure a table, and its grid comes back mostly blank. */
export const MIN_FILLED_CELLS = 0.5;

/** A GitHub-flavoured Markdown table; the first row is the header. Null when the grid is not a
 * table after all — one row, one column, or mostly empty — in which case the lines stand. */
export function tableToMarkdown(rows: readonly (readonly string[])[]): string | null {
  const width = Math.max(0, ...rows.map((row) => row.length));
  if (rows.length < 2 || width < 2) return null;
  const filled = rows.flat().filter((cell) => cell.trim() !== "").length;
  if (filled < MIN_FILLED_CELLS * rows.length * width) return null;
  const padded = rows.map((row) => [...row, ...Array<string>(width - row.length).fill("")]);
  const line = (cells: readonly string[]) => `| ${cells.map(escapeCell).join(" | ")} |`;
  const [header = [], ...body] = padded;
  return [
    line(header),
    `| ${Array<string>(width).fill("---").join(" | ")} |`,
    ...body.map(line),
  ].join("\n");
}

interface Piece {
  box: OcrBox;
  text: string;
}

function blockText(block: OcrBlock): string | null {
  if (block.kind === "formula") {
    const latex = block.latex.trim();
    return latex === "" ? null : `$$${latex}$$`;
  }
  return tableToMarkdown(block.rows);
}

/**
 * The page as lines of text in reading order, blocks written in place of the lines they own.
 * A block's text is surrounded by blank lines so it stays a paragraph of its own downstream.
 */
export function composeRecognizedPage(page: OcrPageResult): string[] {
  const owned = new Set<OcrLine>();
  const pieces: Piece[] = [];
  for (const block of page.blocks) {
    const lines = page.lines.filter(
      (line) => !owned.has(line) && insideFraction(line.box, block.box) >= OWNED_FRACTION,
    );
    // A formula that owns no line is one inside a sentence; the text keeps it.
    if (lines.length === 0 && block.kind === "formula") continue;
    const text = blockText(block);
    if (text === null) continue;
    for (const line of lines) owned.add(line);
    pieces.push({ box: block.box, text: `\n${text}\n` });
  }
  for (const line of page.lines) {
    if (!owned.has(line) && line.text.trim() !== "")
      pieces.push({ box: line.box, text: line.text });
  }
  pieces.sort((a, b) => readingOrder(a.box, b.box));
  return pieces.flatMap((piece) => piece.text.split("\n"));
}

File: apps/desktop/src-tauri/src/ocr_page.rs
// Purpose: what `ocr_page` hands back, and the little geometry every stage of reading a page
// shares. A page is lines of text, each with where it sits, plus the blocks that are not
// running text — a table read into rows of cells, a formula read into LaTeX. The shapes here
// are the contract with the TypeScript side (apps/desktop/src/lib/library/ocrPage.ts), which
// composes them into the lines that get indexed; the browser edition's worker answers with
// the same shapes, so that composition is written once.

use oar_ocr_core::processors::BoundingBox;

/// Page pixels, `x0,y0` top-left, `x1,y1` bottom-right.
#[derive(serde::Serialize, Debug, Clone, Copy, PartialEq)]
pub struct PageBox {
    pub x0: f32,
    pub y0: f32,
    pub x1: f32,
    pub y1: f32,
}

impl PageBox {
    pub fn new(x0: f32, y0: f32, x1: f32, y1: f32) -> Self {
        Self { x0, y0, x1, y1 }
    }

    /// The axis-aligned box around a detection polygon.
    pub fn around(polygon: &BoundingBox) -> Self {
        Self::new(
            polygon.x_min(),
            polygon.y_min(),
            polygon.x_max(),
            polygon.y_max(),
        )
    }

    pub fn area(&self) -> f32 {
        (self.x1 - self.x0).max(0.0) * (self.y1 - self.y0).max(0.0)
    }

    /// The share of this box's area that lies inside `outer`, 0 to 1.
    pub fn inside_fraction(&self, outer: &PageBox) -> f32 {
        let own = self.area();
        if own <= 0.0 {
            return 0.0;
        }
        let overlap = PageBox::new(
            self.x0.max(outer.x0),
            self.y0.max(outer.y0),
            self.x1.min(outer.x1),
            self.y1.min(outer.y1),
        )
        .area();
        overlap / own
    }

    pub fn center(&self) -> (f32, f32) {
        ((self.x0 + self.x1) / 2.0, (self.y0 + self.y1) / 2.0)
    }

    /// Grown by `margin` on every side and clamped to a `width` × `height` page, as integers
    /// a crop can be taken with. Empty when the box lies off the page.
    pub fn crop_rect(&self, margin: f32, width: u32, height: u32) -> Option<(u32, u32, u32, u32)> {
        let x0 = (self.x0 - margin).max(0.0) as u32;
        let y0 = (self.y0 - margin).max(0.0) as u32;
        let x1 = ((self.x1 + margin).ceil() as u32).min(width);
        let y1 = ((self.y1 + margin).ceil() as u32).min(height);
        if x1 > x0 && y1 > y0 {
            Some((x0, y0, x1 - x0, y1 - y0))
        } else {
            None
        }
    }
}

/// One recognised line of text, in reading order.
#[derive(serde::Serialize, Debug, Clone, PartialEq)]
pub struct OcrLine {
    pub text: String,
    pub score: f32,
    #[serde(rename = "box")]
    pub page_box: PageBox,
}

/// A region of the page that is not running text, read by the model made for it.
#[derive(serde::Serialize, Debug, Clone, PartialEq)]
#[serde(tag = "kind", rename_all = "lowercase")]
pub enum OcrBlock {
    /// Rows of cells, spans expanded — a spanning cell's text in its first slot, blanks after.
    Table {
        #[serde(rename = "box")]
        page_box: PageBox,
        rows: Vec<Vec<String>>,
    },
    /// LaTeX without delimiters.
    Formula {
        #[serde(rename = "box")]
        page_box: PageBox,
        latex: String,
    },
}

/// Everything read off one page.
#[derive(serde::Serialize, Debug, Clone, PartialEq, Default)]
pub struct OcrPage {
    pub lines: Vec<OcrLine>,
    pub blocks: Vec<OcrBlock>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_inside_fraction_is_the_share_of_the_inner_box_covered() {
        let outer = PageBox::new(0.0, 0.0, 100.0, 100.0);
        assert_eq!(
            PageBox::new(10.0, 10.0, 20.0, 20.0).inside_fraction(&outer),
            1.0
        );
        assert_eq!(
            PageBox::new(90.0, 0.0, 110.0, 10.0).inside_fraction(&outer),
            0.5
        );
        assert_eq!(
            PageBox::new(200.0, 0.0, 210.0, 10.0).inside_fraction(&outer),
            0.0
        );
        assert_eq!(
            PageBox::new(5.0, 5.0, 5.0, 5.0).inside_fraction(&outer),
            0.0
        );
    }

    #[test]
    fn a_crop_rect_grows_by_the_margin_and_stays_on_the_page() {
        let b = PageBox::new(2.0, 3.0, 50.0, 60.0);
        assert_eq!(b.crop_rect(4.0, 52, 100), Some((0, 0, 52, 64)));
        assert_eq!(
            PageBox::new(60.0, 0.0, 70.0, 5.0).crop_rect(0.0, 52, 100),
            None
        );
    }

    #[test]
    fn blocks_serialise_with_their_kind_first() {
        let block = OcrBlock::Formula {
            page_box: PageBox::new(1.0, 2.0, 3.0, 4.0),
            latex: "x^2".into(),
        };
        let json = serde_json::to_value(&block).unwrap();
        assert_eq!(json["kind"], "formula");
        assert_eq!(json["box"]["x1"], 3.0);
        assert_eq!(json["latex"], "x^2");
        let table = OcrBlock::Table {
            page_box: PageBox::new(0.0, 0.0, 1.0, 1.0),
            rows: vec![vec!["a".into(), "b".into()]],
        };
        assert_eq!(serde_json::to_value(&table).unwrap()["rows"][0][1], "b");
    }
}
